kmk/kmk_keyboard.py

```diff
@@ -90,223 +87,354 @@ def __repr__(self):
+            ):
+                self._tap_dance_counts[changed_key] = 1
+                self.set_timeout(
+                    self.tap_time, lambda: self._end_tap_dance(changed_key)
+                )
+                self._tapping = True
+            else:
+                self._tap_dance_counts[changed_key] += 1
+
+            if changed_key not in self._tap_side_effects:
+                self._tap_side_effects[changed_key] = None
+        else:
+            has_side_effects = self._tap_side_effects[changed_key] is not None
+            hit_max_defined_taps = self._tap_dance_counts[changed_key] == len(
+                changed_key.codes
+            )
+
+            if has_side_effects or hit_max_defined_taps:
+                self._end_tap_dance(changed_key)
+
+        return self
+
+    def _end_tap_dance(self, td_key):
+        v = self._tap_dance_counts[td_key] - 1
+
+        if v >= 0:
+            if td_key in self.keys_pressed:
+                key_to_press = td_key.codes[v]
+                self.add_key(key_to_press)
+                self._tap_side_effects[td_key] = key_to_press
+                self.hid_pending = True
+            else:
+                if self._tap_side_effects[td_key]:
+                    self.remove_key(self._tap_side_effects[td_key])
+                    self._tap_side_effects[td_key] = None
+                    self.hid_pending = True
+                    self._cleanup_tap_dance(td_key)
+                else:
+                    self.tap_key(td_key.codes[v])
+                    self._cleanup_tap_dance(td_key)
 
-    def go(self, hid_type=HIDModes.USB, **kwargs):
+        return self
+
+    def _cleanup_tap_dance(self, td_key):
+        self._tap_dance_counts[td_key] = 0
+        self._tapping = any(count > 0 for count in self._tap_dance_counts.values())
+        return self
+
+    def set_timeout(self, after_ticks, callback):
+        if after_ticks is False:
+            # We allow passing False as an implicit "run this on the next process timeouts cycle"
+            timeout_key = ticks_ms()
+        else:
+            timeout_key = ticks_ms() + after_ticks
+
+        while timeout_key in self._timeouts:
+            timeout_key += 1
+
+        self._timeouts[timeout_key] = callback
+        return timeout_key
+
+    def _cancel_timeout(self, timeout_key):
+        if timeout_key in self._timeouts:
+            del self._timeouts[timeout_key]
+
+    def _process_timeouts(self):
+        if not self._timeouts:
+            return self
+
+        current_time = ticks_ms()
+
+        # cast this to a tuple to ensure that if a callback itself sets
+        # timeouts, we do not handle them on the current cycle
+        timeouts = tuple(self._timeouts.items())
+
+        for k, v in timeouts:
+            if k <= current_time:
+                v()
+                del self._timeouts[k]
+
+        return self
+
+    #####
+    # SPLICE END: INTERNAL STATE
+    # TODO FIXME REMOVE THIS
```

Comment:
same remarks as above

Reply:
That one is on you. I couldn't figure out how to get this one out cleanly, which is why it's there. would mark this one as help_required.

Reply:
at this point you should be able to remove the lines assuming the splicing work is done, right? just add another commit to the stack and make sure it's at the end of the rebase and not early on?

Reply:
I was under the impression that these functions were still used see calls to them.

Reply:
Totally need sleep. Left these is to review as I didn't know if something else needed to get out. Intended to remove but wanted to leave markers for review reference.